### PR TITLE
WIP: feat(cursor): new cursor.transformStream method

### PR DESCRIPTION
**Pull request by rweinberger**
Created a new method, cursor.transformStream, which
returns a readable stream with the transformation
applied. Does not mutate the cursor.

Fixes NODE-1208

### CODE REVIEW

File: lib/cursor.js [hunk 1]
```diff
@@ -1123,6 +1125,30 @@ Cursor.prototype.stream = function(options) {
   return this;
 };
 
+/**
+ * Return a modified Readable stream that applies a given transform function, if supplied. If none supplied,
+ * returns a stream of unmodified docs.
+ * @method
+ * @param {object} [options=null] Optional settings.
+ * @param {function} [options.transform=null] A transformation method applied to each document emitted by the stream.
```
Comment (mbroadst) [inline lib/cursor.js]: no need to add the `=null` to optional parameters

File: test/functional/cursor_tests.js [hunk 2]
```diff
@@ -4423,4 +4423,79 @@ describe('Cursor', function() {
...
+          stream.once('end', function() {
+            expect(resultSet).to.deep.equal(expectedSet);
+            cursor.close(() => client.close(() => done()));
+          });
+
+          stream.once('error', function(err) {
+            cursor.close(() => client.close(() => done(err)));
+          });
+
+          stream.resume();
```
Comment (mbroadst) [inline test/functional/cursor_tests.js]: You shouldn't need to call `stream.resume()` here. When you register the listener for `data` above, it automatically triggers the command + getMore

File: test/functional/cursor_tests.js [hunk 3]
```diff
@@ -4423,4 +4423,79 @@ describe('Cursor', function() {
...
+        .then(_cursor => (cursor = _cursor))
+        .then(() => cursor.transformStream(transformParam))
+        .then(stream => {
+          stream.on('data', function(doc) {
+            resultSet.add(doc);
+          });
+
+          stream.once('end', function() {
+            expect(resultSet).to.deep.equal(expectedSet);
+            cursor.close(() => client.close(() => done()));
```
Comment (mbroadst) [inline test/functional/cursor_tests.js]: it looks like this code can be defined commonly as the `cleanup` handler or something, it's called three times in this test.

File: test/functional/cursor_tests.js [hunk 4]
```diff
@@ -4423,4 +4423,79 @@ describe('Cursor', function() {
...
+  it('transformStream should apply the supplied transformation function to each document in the stream', function(done) {
+    const configuration = this.configuration;
+    const client = configuration.newClient({ w: 1 }, { poolSize: 1, auto_reconnect: false });
+    const expectedDocs = [{ _id: 0, b: 1, c: 0 }, { _id: 1, b: 1, c: 0 }, { _id: 2, b: 1, c: 0 }];
+
+    testTransformStream(
+      client,
+      configuration,
+      'transformStream-test-transform',
+      doc => {
```
Comment (mbroadst) [inline test/functional/cursor_tests.js]: FYI this can be rewritten as: `doc => ({ _id: doc._id, b: doc.a.b, c: doc.a.c })` (note the wrapping parens)

**Comment by daprahamian** [comment]
@mbroadst The "breaking" aspect is that we originally wanted this function to replace `.stream`, b/c `stream`'s behavior doesn't really make sense. However, we did not want to make a breaking change.

**Comment by mbroadst** [comment]
@daprahamian ah! I see, and agree. Let's maybe leave a note in here to remind ourselves to replace `stream` with `transformStream` in the next major release?

File: lib/cursor.js [hunk 5]
```diff
@@ -1123,6 +1125,30 @@ Cursor.prototype.stream = function(options) {
...
+ * @return {stream}
+ */
+Cursor.prototype.transformStream = function(options) {
+  let streamOptions = options || {};
```
Comment (daprahamian) [inline lib/cursor.js]: `const`

File: lib/cursor.js [hunk 6]
```diff
@@ -1123,6 +1125,30 @@ Cursor.prototype.stream = function(options) {
...
+  if (streamOptions.transform != null && typeof streamOptions.transform === 'function') {
```
Comment (daprahamian) [inline lib/cursor.js]: the `streamOptions.transform != null` is superfluous here. You can just do

```js
if (typeof streamOptions.transform === 'function') {
```

File: test/functional/cursor_tests.js [hunk 7]
```diff
@@ -4423,4 +4423,79 @@ describe('Cursor', function() {
       });
     });
   });
+
+  function testTransformStream(client, configuration, dbName, transformFunc, expectedSet, done) {
```
Comment (daprahamian) [inline test/functional/cursor_tests.js]: 1. I really like you making this single utility function.
2. You are using `dbName` here as a collection name. You should rename it.
3. This is a lot of parameters for a function call. I'd replace these with just two:

```js
function testTransformStream(config, done) {
```

And you can stick everything you need in `config`

**Pull request merged**